server/api/public.py

```diff
@@ -42,15 +47,108 @@
 # Specifying election results in a query param could cause us to hit URL size limits
 @api.route("/public/sample-sizes", methods=["POST"])
 def public_compute_sample_sizes():
-    compute_sample_sizes_input = request.get_json()
-    validate(compute_sample_sizes_input, COMPUTE_SAMPLE_SIZES_INPUT_SCHEMA)
-    sample_sizes = {
-        "ballotComparison": {},
-        "ballotPolling": {},
-        "batchComparison": {},
+    contest_for_sampler = parse_compute_sample_sizes_input(request.get_json())
+
+    audit_types = ["ballotComparison", "ballotPolling", "batchComparison"]
+    sample_sizes = {audit_type: {} for audit_type in audit_types}
+    sample_size_functions = {
+        "ballotComparison": compute_ballot_comparison_sample_size,
+        "ballotPolling": compute_ballot_polling_sample_size,
+        "batchComparison": compute_batch_comparison_sample_size,
     }
-    for risk_limit_percentage in list(range(0, 21)):
-        sample_sizes["ballotComparison"][str(risk_limit_percentage)] = 0
-        sample_sizes["ballotPolling"][str(risk_limit_percentage)] = 0
-        sample_sizes["batchComparison"][str(risk_limit_percentage)] = 0
+    for audit_type in audit_types:
+        for risk_limit_percentage in list(range(0, 21)):
+            sample_size_function = sample_size_functions[audit_type]
+            sample_sizes[audit_type][str(risk_limit_percentage)] = sample_size_function(
+                contest_for_sampler, risk_limit_percentage
+            )
+
     return jsonify(sample_sizes)
+
+
+def parse_compute_sample_sizes_input(
+    compute_sample_sizes_input: Any,
+) -> sampler_contest.Contest:
+    validate(compute_sample_sizes_input, COMPUTE_SAMPLE_SIZES_INPUT_SCHEMA)
+
+    election_results = compute_sample_sizes_input["electionResults"]
+    candidates = election_results["candidates"]
+    num_winners = election_results["numWinners"]
+    total_ballots_cast = election_results["totalBallotsCast"]
+
+    if all(candidate["votes"] == 0 for candidate in candidates):
+        raise Conflict("At least 1 candidate must have greater than 0 votes")
+
+    if num_winners >= len(candidates):
+        raise Conflict("Number of winners must be less than number of candidates")
+
+    candidate_votes_sum = sum(candidate["votes"] for candidate in candidates)
+    if total_ballots_cast < candidate_votes_sum:
+        raise Conflict("Total ballots cast cannot be less than sum of candidate votes")
+
+    contest = Contest(
+        choices=[
+            ContestChoice(
+                id=f"candidate-{i}",
+                num_votes=candidate["votes"]
+                + 0.01  # Safeguard against math errors involving 0s
+                if candidate["votes"] == 0
+                else candidate["votes"],
+            )
+            for i, candidate in enumerate(candidates)
+        ],
+        num_winners=num_winners,
+        total_ballots_cast=total_ballots_cast,
+        votes_allowed=num_winners,
```

Comment:
Related to the discussion about number of votes allowed (https://github.com/votingworks/arlo/pull/1600#discussion_r982681006), because we don't expose an input for this, I'm just setting number of votes allowed to number of winners.

I've done a quick spot check to verify that the audit math doesn't actually need number of votes allowed.

Reply:
Dropped a question in Slack for greater visibility: https://votingworks.slack.com/archives/CKCVA0F9S/p1664387582424249